mlflow/java/spark/src/main/scala/org/mlflow/spark/autologging/MlflowAutologEventPublisher.scala

```diff
@@ -0,0 +1,130 @@
+  */
+object MlflowAutologEventPublisher extends MlflowAutologEventPublisherImpl {
+
+}
+
+/**
+ * Trait implementing a publisher interface for publishing events on Spark datasource reads to
+ * a set of listeners. See the design doc:
+ * https://docs.google.com/document/d/11nhwZtj-rps0stxuIioFBM9lkvIh_ua45cAFy_PqdHU/edit for more
+ * details.
+ */
+private[autologging] trait MlflowAutologEventPublisherImpl {
+  private val logger = LoggerFactory.getLogger(getClass)
+
+  private[autologging] var sparkQueryListener: SparkDataSourceListener = _
+  private val ex = new ScheduledThreadPoolExecutor(1)
+  private[autologging] var subscribers: mutable.LinkedHashMap[String, MlflowAutologEventSubscriber] =
+    mutable.LinkedHashMap[String, MlflowAutologEventSubscriber]()
+  private var scheduledTask: ScheduledFuture[_] = _
+
+  def spark: SparkSession = {
+    SparkSession.builder.getOrCreate()
+  }
+
+  // Exposed for testing
+  private[autologging] def getSparkDataSourceListener: SparkDataSourceListener = {
+    new SparkDataSourceListener()
+  }
+
+  // Initialize Spark listener that pulls Delta query plan information & bubbles it up to registered
+  // Python subscribers, along with a GC loop for removing unrespoins
+  def init(gcDeadSubscribersIntervalSec: Int = 1): Unit = synchronized {
+    if (sparkQueryListener == null) {
+      val listener = getSparkDataSourceListener
+      // NB: We take care to set the variable only after adding the Spark listener succeeds,
+      // in case listener registration throws. This is defensive - adding a listener should
+      // always succeed.
+      spark.sparkContext.addSparkListener(listener)
+      sparkQueryListener = listener
+      // Schedule regular cleanup of detached subscribers, e.g. those associated with detached
+      // notebooks
+      val task = new Runnable {
+        def run(): Unit = {
+          unregisterBrokenSubscribers()
+        }
+      }
+      scheduledTask = ex.scheduleAtFixedRate(task, 1, gcDeadSubscribersIntervalSec, TimeUnit.SECONDS)
+    }
+  }
+
+  def stop(): Unit = synchronized {
+    if (sparkQueryListener != null) {
+      spark.sparkContext.removeSparkListener(sparkQueryListener)
+      sparkQueryListener = null
+      while(!scheduledTask.cancel(false)) {
+        Thread.sleep(1000)
+      }
+      subscribers = mutable.LinkedHashMap()
+    }
+  }
+
+  def register(subscriber: MlflowAutologEventSubscriber): String = synchronized {
+    if (sparkQueryListener == null) {
+      throw new RuntimeException("Please call init() before attempting to register a subscriber")
+    }
+    this.synchronized {
+      val uuid = java.util.UUID.randomUUID().toString
+      subscribers.put(uuid, subscriber)
+      uuid
+    }
+  }
+
+  /** Unregister subscribers broken e.g. due to detaching of the associated Python REPL */
+  private[autologging] def unregisterBrokenSubscribers(): Unit = synchronized {
+    val brokenUuids = subscribers.flatMap { case (uuid, listener) =>
```

Comment:
Do wee need to do the pings under the lock?

Reply:
Hm I don't think we do, I'll confirm this w/ you tomorrow.

In particular we need to:
1. Get a snapshot iterator `it` over the elements of `subscribers` at some point in time
2. Iterate over `it` & ping each subscriber, get a list of bad repl IDs
3. Remove each bad repl ID from our map

Steps 2 & 3 can definitely be done outside of a lock (replIds are unique, so we won't ever accidentally remove a working REPL in step 3). Step 1 we can achieve by making a copy of `subscribers` (e.g. via `subscribers.toSeq`). The [Scala API docs](https://www.scala-lang.org/api/2.12.0/scala/collection/mutable/LinkedHashMap.html#toSeq:Seq[(K,V)]) don't guarantee that `toSeq` is threadsafe, so maybe we should synchronize this part at least.

Reply:
That being said, we are assuming here that there's only one thread ever running `unregisterBrokenSubscribers`, so should at least add a comment stating that assumption when we create the GC thread/in this method docstring